### PR TITLE
Issue #5586: Drop Cobertura

### PR DESCRIPTION
**Pull request by pbludov**
Issue #5586

- cobertura replaced with jacoco where possible
- Jacoco check extracted from it's profile
- JDK9 build with tests
- Sonarqube coverage

### CODE REVIEW

**Comment by pbludov** [comment]
@romani I need an advice on how to solve the Travis CI issue.

**Comment by rnveach** [comment]
This is more than just a 1 line change.
We need to remove cobertura completely.
See everything listed in: https://github.com/checkstyle/checkstyle/search?utf8=%E2%9C%93&q=cobertura&type=

**Comment by rnveach** [comment]
@pbludov 
> [INFO] --- cobertura-maven-plugin:2.7:cobertura (default-cli) @ checkstyle ---

cobertura shouldn't be running at all in travis. We are dropping it for jacoco. Like I stated above, everything cobertura related has to be removed/changed.

**Comment by romani** [comment]
@pbludov , please let us know if you still need help with travis.

**Comment by pbludov** [comment]
@romani  No, thank you. I nearly know what needs to be done.
@rnveach Yes, I understood that it is complicated issue. My goal is to minimize the number of changed files to enable the JDK9 build. All the rest will be in separated PR.

**Comment by pbludov** [comment]
@romani  I've switched Codecov to use the Jacoco report. But I have no idea on what to do with sonarqube.

**Comment by romani** [comment]
sonarqube is triggered only from master build,  we can skip it for now, it does not do any enforcing validation, their website does not allow customization now, but they plan to add it in future

**Comment by pbludov** [comment]
@romani about sonarcube. According to the docs, only one line is need:
```
 <sonar.jacoco.reportPaths>${project.build.directory}/jacoco/jacoco.exec</sonar.jacoco.reportPaths>
```
Sample: https://github.com/maxkarelov/sonarqube-jacoco-maven-sample/blob/master/pom.xml
Docs: https://docs.sonarqube.org/display/PLUG/Code+Coverage+by+Unit+Tests+for+Java+Project

**Comment by romani** [comment]
@pbludov , please rebase this PR

**Comment by pbludov** [comment]
Done.

File: pom.xml [hunk 1]
```diff
@@ -751,6 +747,171 @@
...
+            <id>default-restore-instrumented-classes</id>
+            <goals>
+              <goal>restore-instrumented-classes</goal>
+            </goals>
+          </execution>
+          <execution>
+            <id>default-report</id>
+            <phase>site</phase>
+            <goals>
+              <goal>report</goal>
```
Comment (romani) [inline pom.xml]: please do:
```
$ git diff
diff --git a/pom.xml b/pom.xml
index dda176c..b83192e 100644
--- a/pom.xml
+++ b/pom.xml
@@ -768,10 +768,13 @@
           </execution>
           <execution>
             <id>default-report</id>
-            <phase>site</phase>
+            <phase>verify</phase>
             <goals>
               <goal>report</goal>
             </goals>
+            <configuration>
+              <dataFile>${project.build.directory}/jacoco/jacoco.exec</dataFile>
+            </configuration>
```

it is better to generate report on verify phase, to have report available right after verification is failed. We ask user to use "mvn verify" for all verifications. Without html report it is impossible to get to know what to fix. All other plugins report to terminal output some report that is enough to fix problems.

File: pom.xml [hunk 2]
```diff
@@ -751,6 +747,171 @@
...
+            </goals>
+          </execution>
+          <execution>
+            <id>default-check</id>
+            <goals>
+              <goal>check</goal>
+            </goals>
+            <configuration>
+              <dataFile>${project.build.directory}/jacoco/jacoco.exec</dataFile>
```
Comment (romani) [inline pom.xml]: let make variable for `${project.build.directory}/jacoco/jacoco.exec` to reuse it.

http://www.eclemma.org/jacoco/trunk/doc/report-mojo.html
>"Default value is: ${project.build.directory}/jacoco.exec."

default value for path is different, why to use non default ?

File: pom.xml [hunk 3]
```diff
@@ -1345,7 +1492,7 @@
           -->
           <excludedPages>
             <excludedPage>dependencies.html</excludedPage>
-            <excludedPage>cobertura/**</excludedPage>
+            <excludedPage>jacoco/**</excludedPage>
```
Comment (romani) [inline pom.xml]: confirmed, as plugin generate link that is not 100% valid, links check gives `http://www.jacoco.org/jacoco: 301 Moved Permanently` . NO `/` at the end of url.

File: .travis.yml [hunk 4]
```diff
@@ -187,7 +181,7 @@ matrix:
     - jdk: oraclejdk9
       env:
         - DESC="build with JDK9"
-        - CMD="mvn -e package -Passembly"
+        - CMD="mvn -e package"
```
Comment (romani) [inline .travis.yml]: what is a reason ?

**Comment by pbludov** [comment]
Rebased again after b058aeed964ed9e6ea7d74de3b5870ef99acff49

File: .travis.yml [hunk 5]
```diff
@@ -39,19 +39,13 @@ matrix:
...
-        - DESC="jacoco"
-        - CMD="export MAVEN_OPTS='-Xmx2000m' && mvn -e clean verify -Pno-validations -Pjacoco-check -DskipTests=false"
-
-    # cobertura and codecov (oraclejdk8)
-    - jdk: oraclejdk8
-      env:
-        - DESC="cobertura and codecov"
-        - CMD="./.ci/travis/travis.sh cobertura-check"
+        - DESC="jacoco and codecov"
+        - CMD="export MAVEN_OPTS='-Xmx2000m' && mvn -e clean site -Pno-validations -Djacoco.skip=false -DskipTests=false"
```
Comment (romani) [inline .travis.yml]: looks over complicated ..... 
thee should be smth like `mvn test jacoco:report` .... but for some reason I can not make it work

Comment (pbludov) [reply .travis.yml]: Me neither. I think the point is that JaCoCo need to make the instumentation, do the tests, then create a report. Here that almost works:
```
mvn -e clean test jacoco:instrument@default-instrument jacoco:check@default-check jacoco:report@default-report
```
but this is even worse.

Comment (romani) [reply .travis.yml]: @pbludov , please resolve all other points, we will think about it later.

File: .travis.yml (hunk 4 above)
Comment (pbludov) [reply .travis.yml]: To make sure that all the tests have passed. `-Passembly` was a workaround at a time when Powermock was incompatible with JDK9. Now it is even possible to build the `site` phase.
I'm going to carefully check the compatibility with JDK9, after completing this task.

File: pom.xml (hunk 2 above)
Comment (pbludov) [reply pom.xml]: I dont know. I took the existing commit and developed it a little.
Using the default value seems to solve a lot of problems.

File: pom.xml (hunk 1 above)
Comment (pbludov) [reply pom.xml]: done

File: .travis.yml [hunk 6]
```diff
@@ -186,8 +180,8 @@ matrix:
     # JDK9 build
     - jdk: oraclejdk9
       env:
-        - DESC="build with JDK9"
-        - CMD="mvn -e package -Passembly"
+        - DESC="JDK9 site, assembly"
+        - CMD="mvn -e site -Dlinkcheck.skip=true && mvn -e package -Passembly"
```
Comment (romani) [inline .travis.yml]: please move it to separate commit(with reference jdk9 issue) or to another PR.
cobertura --> jacoco is unrelated to jdk8 --> jdk9 . Even they a bit bound.

Comment (pbludov) [reply .travis.yml]: done. Reverted to original `CMD="mvn -e package -Passembly"`

File: .travis.yml [hunk 7]
```diff
@@ -39,19 +39,13 @@ matrix:
...
-        - CMD="export MAVEN_OPTS='-Xmx2000m' && mvn -e clean verify -Pno-validations -Pjacoco-check -DskipTests=false"
-
-    # cobertura and codecov (oraclejdk8)
-    - jdk: oraclejdk8
-      env:
-        - DESC="cobertura and codecov"
-        - CMD="./.ci/travis/travis.sh cobertura-check"
+        - DESC="jacoco and codecov"
+        - CMD="export MAVEN_OPTS='-Xmx2000m' && mvn -e clean verify"
         - CMD_AFTER_SUCCESS="bash <(curl -s https://codecov.io/bash)"
```
Comment (romani) [inline .travis.yml]: Verify phase will run all verification tools (checkstyle, pm, finbugs,.....) It will take time. But here we want to run only coverage check, please use '-Pno-validations....' as you did before. One day we will find way to run jacoco only.

Comment (pbludov) [reply .travis.yml]: I've found the minimalistic version:
```
mvn clean test jacoco:restore-instrumented-classes jacoco:report@default-report
```

File: pom.xml (hunk 1 above)
Comment (pbludov) [reply pom.xml]: reverted back to the `site` phase since `verify` is no more required to get the JaCoCo report.

File: src/xdocs/contributing.xml [hunk 8]
```diff
@@ -83,15 +83,15 @@
...
-        <a href="http://cobertura.github.io/cobertura/">Cobertura</a>.
+        <a href="http://www.eclemma.org/jacoco/">JaCoCo</a>.
 
         This means it is not sufficient to pass all tests, but the tests
         should ideally execute each line in the code, code coverage should be 100%. To generate the
-        Cobertura report, run the Maven command
+        JaCoCo report, run the Maven command
 
-        <code>mvn cobertura:cobertura</code>.
+        <code>mvn test jacoco:restore-instrumented-classes jacoco:report@default-report</code>.
```
Comment (pbludov) [inline src/xdocs/contributing.xml]: Still not ideal. Perhaps some bash script may help.

Comment (romani) [reply src/xdocs/contributing.xml]: it is more then enough for now. We will improve it some time later on.

File: pom.xml [hunk 9]
```diff
@@ -751,6 +746,184 @@
...
+                <!-- Parser related classes -->
+                <exclude>com/puppycrawl/tools/checkstyle/grammars/*.class</exclude>
+                <exclude>com/puppycrawl/tools/checkstyle/grammars/javadoc/*.class</exclude>
+                <exclude>com/puppycrawl/tools/checkstyle/JavadocDetailNodeParser$JavadocParserErrorStrategy.class</exclude>
+                <!-- Swing related classes -->
+                <exclude>com/puppycrawl/tools/checkstyle/gui/*.class</exclude>
+                <!-- Deprecated classes -->
+                <exclude>com/puppycrawl/tools/checkstyle/checks/naming/AbstractTypeParameterNameCheck.class</exclude>
+                <!-- Seems to be a bug in JaCoCo. It complains about static init on a compiler generated class. -->
+                <exclude>com/puppycrawl/tools/checkstyle/checks/annotation/AnnotationUseStyleCheck$1.class</exclude>
```
Comment (rnveach) [inline pom.xml]: Has the issue been reported to Jacoco?

File: pom.xml [hunk 10]
```diff
@@ -955,7 +1128,7 @@
         <version>2.20.1</version>
         <configuration>
           <systemPropertyVariables>
-            <jacoco-agent.destfile>${project.build.directory}/jacoco/jacoco.exec</jacoco-agent.destfile>
+            <jacoco-agent.destfile>${project.build.directory}/jacoco.exec</jacoco-agent.destfile>
```
Comment (rnveach) [inline pom.xml]: why the change?

Comment (romani) [reply pom.xml]: just to use default path of jacoco plugin, it was my request.

File: pom.xml (hunk 9 above)
Comment (romani) [reply pom.xml]: it might be hard to open good issue for this...

@pbludov , please create some issue with uncertain description, we will use link here.
After merge, steps to reproduce in jacoco issue will be simple: clone, remove exclude, run command.

In the same time we will have some tracking number here for future to recheck.
Comment should be like "till https:///////#XXXXX"

Comment (pbludov) [reply pom.xml]: This may be known issue, I'll take a look.

Comment (pbludov) [reply pom.xml]: No, there is no issues like that in the Jacoco tracker.
The problem is easy to reproduce, so I created a new task:
https://github.com/jacoco/jacoco/issues/660

Comment (romani) [reply pom.xml]: Please reference it in comment above suppression

Comment (rnveach) [reply pom.xml]: Thanks.

Comment (rnveach) [reply pom.xml]: @pbludov @romani jacoco/jacoco#660 is now closed and has a milestone of 0.8.2 .
